### ISSUE TITLE
Grid field actions to move into a more options dropdown

### ISSUE THREAD

**Issue by clarkepaul** [silverstripe/silverstripe-admin#287]
# Overview

Now that grid field items have more actions (including publish, unpublish, archive, and potentially more like Add to campaign) we need a more flexible approach to showing available actions on the grid field holder itself and not just in the detailed view of items. We also give too much prominence to the delete action on grid field items—we should be encouraging constructive actions over deconstructive actions. 

This is the proposed functionality for grid items and how content blocks (WIP) are shaping up to use the same functionality.
![pl_gridfield_options](https://user-images.githubusercontent.com/555033/31205067-9e99911e-a9cb-11e7-9cf6-d9e182953836.png)

Side note: I was sure there was an issue for this already but can't find it so creating a this as a new one (might have been part of another issue but excluded from AC's)?

## Acceptance Criteria

 * Developers can use the GridField PHP API to add options
 * Options can be global or row specific
 * Options can be grouped (with a visual separator)
 * Existing GridFields (ModelAdmin, Security, etc) use this new component
 * The dev documentation for setting up GridFields reflects this change
 * GridField still supports multiple buttons in parallel to this new UI component

## Notes

 * Excludes list in AssetAdmin
 * This information will need to come through as structured data for React-based GridFields soon
 * Auto-adding tabs from detail forms is out of scope
 * Should be built as a React component embedded in a Entwine-based GridField (so we don't have to rewrite it later)
 * Track under ORB-175

## Pull requests
- [x] https://github.com/silverstripe/silverstripe-framework/pull/8083
- [x] https://github.com/silverstripe/silverstripe-admin/pull/530
---
- (OLD REFERENCE ONLY) https://github.com/symbiote/silverstripe-gridfieldextensions/pull/225

**Comment by chillu** [comment]
We're past the point where we can add new features in 4.0, sorry - removing milestone.

**Comment by chillu** [comment]
@clarkepaul Does this make it clear enough that clicking on the whole row triggers the edit view on a standard GridField? Do you expect mixed cases where you have primary actions shown alongside a "more actions" button?

**Comment by clarkepaul** [comment]
There is a cursor change and a hover state on gridfield items, and if we make it a consistent pattern throughout I don't think its going to be an issue (same as gmail/gdrive and a lot of other UI's). If we do find its going to be an issue (through testing) then we can add an action to the options menu for "Edit". I'm pretty confident it will be okay as long the pattern is carried out everywhere through the UI, eg campaign items, history items, search lists all need to be selectable by clicking on the row.

**Comment by chillu** [comment]
Pull request at https://github.com/symbiote/silverstripe-gridfieldextensions/pull/225 and https://github.com/dnadesign/silverstripe-elemental/pull/148

**Comment by flamerohr** [comment]
Added https://github.com/silverstripe/silverstripe-admin/pull/459 to fix the PopoverField accessibility issue

**Comment by clarkepaul** [comment]
Not sure if this was missed but Bootstrap dropdowns should be used for places where actions are rather than popovers (popovers for text, dropdowns for action list), I have mentioned this many times but could easily be overlooked with so many places this has been discussed see https://github.com/symbiote/silverstripe-gridfieldextensions/pull/225#issuecomment-344734543
I presume dropdowns are now available with ReactStrap @flamerohr for such UI's?

**Comment by chillu** [comment]
Talked to @flamerohr, and he's confident that we can achieve the same accessibility behaviours with a popover, because it's easier to do (we already have a component for it). I would've preferred a dropdown UI control as well, but in the end as long as focus management, keyboard access and ARIA metadata is in place, I don't mind.

**Comment by clarkepaul** [comment]
No problem, I was hoping the dropdown component would magically be there with ReactStrap but thats wishful thinking :). Hopefully this won't add too much refactoring later as there is going to loads of places its going to be used, visually they don't differ to much so not too concerned about that part.

**Comment by flamerohr** [comment]
WIP on this branch: https://github.com/open-sausages/silverstripe-framework/tree/pulls/4/action-menu-heroes--assemble

So far I've built the gridfield component to cater for non-react and react behaviours, the component generates a type of schema for the react component to be able to use and render with - but if react is not available it will still be able to fallback to default buttons-without-a-menu style.

The frontend/react has not been developed, however it should be easy enough to use either Popover or Dropdown - so opting for Dropdown is a good idea, as it's the preferred style.
The entwine script to plug the react in is pretty straight-forward like how entwine-react plugs we have in core.
